mono/mini/method-to-ir.c

```diff
@@ -8560,24 +8583,43 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				}
 
 				if (info_type == MONO_PATCH_INFO_ICALL_ADDR) {
+					// FIXME tailcall?
+					tailcall_calli = FALSE;
 					ins = (MonoInst*)mono_emit_abs_call (cfg, MONO_PATCH_INFO_ICALL_ADDR_CALL, info_data, fsig, sp);
 					NULLIFY_INS (addr);
 					goto calli_end;
 				} else if (info_type == MONO_PATCH_INFO_JIT_ICALL_ADDR) {
+					// FIXME tailcall?
+					tailcall_calli = FALSE;
 					ins = (MonoInst*)mono_emit_abs_call (cfg, info_type, info_data, fsig, sp);
 					NULLIFY_INS (addr);
 					goto calli_end;
 				}
 			}
-			ins = (MonoInst*)mini_emit_calli (cfg, fsig, sp, addr, NULL, NULL);
+			ins = (MonoInst*)mini_emit_calli_full (cfg, fsig, sp, addr, NULL, NULL, tailcall_calli);
 
 			calli_end:
 			if (ins_flag & MONO_INST_TAILCALL)
-				test_tailcall (cfg, FALSE); // Currently we never tailcall calli.
+				test_tailcall (cfg, tailcall_calli);
+
+			// Tailcall does not return, and the subsequent code confuses the JIT, and
+			// can usually be optimized away. However note that valid code can branch
```

Comment:
What is the problem with having code past the tailcall ? For the rest of the JIT, a tailcall looks just like a normal call.

Reply:
Whenever I don't do this, I get an assertion failure. A register is -1.
Perhaps the problem is when only doing half the work though.

Reply:
I  could just reduce the commentary and leave it looking like a nice little optimization. I can reproduce the assertion failure.

Reply:
Actually..they don't entirely go together. There is a reason to have two variables.
I am missing the test case for why they are two variables -- because you can branch around the call to ret. Anyway, there was always an assertion failure when I didn't "finish" this, that I didn't try to fix/understand otherwise.

Reply:
I meant -- I can try to reproduce the failure.
It was possibly
				g_assert (vreg != -1);
in
mono_handle_global_vregs but I'm guessing from memory.

Reply:
One solution is to just generate the same code, i.e. dont skip ret, etc. It will lead to some extra code generated, but keep the JIT simpler.

Reply:
I'm pretty sure if I do nothing, leave the push and the ret, it failed an assert. I'd have to disable  the code and  rerun the test to be sure. It is a little wierd when you look at the generated code and see the obviously dead code, but maybe only 3 or so of us look and maybe only 1 thinks it wierd.

The optimization is based on regular call/tailcall of course, I didn't come up with it.

Reply:
I'll try it. Not worth so much talk.

Reply:
Wild guess while I wait for tests..
```
mono_emit_call_args
	if (tailcall) {
...
	} else if (!MONO_TYPE_IS_VOID (sig_ret))
		call->inst.dreg = alloc_dreg (cfg, (MonoStackType)call->inst.type);
```

and then you ret after that, probably not happy.. so, it is fixable, if you want tailcalls to look more like calls. They'd still be saving stack which is the point.

I also suspect the branch around call doesn't work.
Though maybe only allowed for void functions??
i.e. I don't know what is verifiable, or ECMA allows for verification to be easier.

Reply:
I'm not seeing the failure now with a little local testing.
Want me to remove all  skip_ret/push_reslogic??
It does seem suspicious if removed.
From call/callvirt too?

Another option, really, is to share more code between calli and regular  call/callvirt.
How about that? I meant to in the first place but waffled.  Let me try..

Reply:
I think we should try to remove that logic if possible, it will lead to slightly bigger generated code, but the JIT code will be simpler and less error prone.

Reply:
I don't think this is all that complicated, but it is subjective.
Let me try to have one copy instead of two, and then we can consider zero or one.
Two is harder to justify than one or zero.